addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -3225,10 +3225,15 @@ export class OdooEditor extends EventTarget {
         const files = getImageFiles(ev.clipboardData);
         const targetSupportsHtmlContent = isHtmlContentSupported(sel.anchorNode);
         const clipboardHtml = ev.clipboardData.getData('text/html');
-        if (files.length && targetSupportsHtmlContent) {
-            this.addImagesFiles(files).then(html => this.execCommand('insertHTML', this._prepareClipboardData(html)));
-        } else if (clipboardHtml && targetSupportsHtmlContent) {
-            this.execCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
+        if ((files.length || clipboardHtml) && targetSupportsHtmlContent) {
+            const clipboardDiv = document.createElement('div');
+            clipboardDiv.innerHTML = clipboardHtml;
+            const clipboardHtmlIsTable = clipboardDiv.getElementsByTagName('table').length > 0;
+            if (files.length && !clipboardHtmlIsTable) {
+                this.addImagesFiles(files).then(html => this.execCommand('insertHTML', this._prepareClipboardData(html)));
+            } else {
+                this.execCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
+            }
```

Comment:
A few remarks:
- Since we have to choose a random tag anyway, I'd use `template` here so that it semantically has no additional meaning.
- Let's call `_prepareClipboardData` directly so the html is sanitized and we don't have to call it twice below.
- `getElementsByTagName` is a special method that returns a _live_ list of nodes, which means it follows the state of the tree and update itself automatically when it changs etc. You don't need all that, you only want to know if there is a table. Use `document.querySelector('table')` instead.

Considering all the above, we get:
```js
const clipboard = document.createElement('template');
clipboard.innerHTML = this._prepareClipboardData(html);
if (files.length && !document.querySelector('table') {
    this.addImagesFiles(files).then(html => this.execCommand('insertHTML', clipboard.content));
} else {
    this.execCommand('insertHTML', clipboard.content);
}
```

Reply:
Ah, nice catch, I didn't notice `addImagesFiles` actually returned new html. It's very poorly named, and its documentation is wrong even, but anyway... The code then becomes:
```js
const clipboard = document.createElement('template');
clipboard.innerHTML = this._prepareClipboardData(clipboardHtml);
if (files.length && !document.querySelector('table') {
    this.addImagesFiles(files).then(html => this.execCommand('insertHTML', html));
} else {
    this.execCommand('insertHTML', clipboard.content);
}
```

Reply:
@dmo-odoo I assume I should use clipboard instead of document in order to replace the `getElementsByTagName` with a `querySelector` ?
I'm also not sure about calling _prepareClipboardData before knowing it's needed, might it make the copy-pasting less responsive for no reason ?

Reply:
I just mean slower.

Reply:
For the `else` case it is exactly the same since we'd end up doing it anyway. For the `if` case it's debatable but the difference is negligible and I'd rather have the `_prepareClipboardData` call for overriding the security ci.

Reply:
Done.
In the `else` I have to use `clipboard.innerHTML` because `clipboard.content` returns a document fragment which is not interpreted as a table but rather as a string in the document.

Reply:
After rebasing, the document fragment is interpreted correctly.
So, I now use `clipboard.content` although I'm not quite sure why we prefer this to `clipboard.innerHTML` since that's what we assigned the result of `_prepareClipboardData` to.